### PR TITLE
switch to watchify(browserify) builder

### PR DESCRIPTION
**Pull request by vvo**
Changes:

- use [watchify](https://github.com/substack/watchify) to wrap and cache [browserify bundles](https://github.com/substack/node-browserify) in both the zuul client and the user entry point.
  This speeds up page refreshs and development cycle when in [local mode](https://github.com/defunctzombie/zuul#running-zuul-locally) with zuul
- switch to [browserify@9](https://github.com/defunctzombie/zuul#running-zuul-locally), this means we need to bump major on zuul
- add some debug around browserify builds timing

### CODE REVIEW

**Comment by vvo** [comment]
Woho, there's another ~500ms to shave on the client bundling.. Let's do that too.

Watchify is really a simple wrapper around browserify, nice API, little code change but I am here to learn what you think about it @defunctzombie.

**Comment by vvo** [comment]
So here's the result when bundling with watchify both zuul client and test files:

2015-02-19-172221_722x289_scrot

Normally it would take at least 3.5 seconds on this little project for tests to be compiled then served to the client.

But now on each pageload/refresh it will take as low as 1 second or maybe less.

What do you think? I really want to provide a good developer workflow, not just CI workflow where we obviously don't care about a 2s shaving on a one time operation.

But the developer workflow is made of refresh/test refresh/test in the browser (with zuul) so, shaving 2 seconds on a small project on every refresh, is great? :)

File: package.json [hunk 1]
```diff
@@ -7,8 +7,8 @@
     "JSON2": "0.1.0",
     "batch": "0.5.0",
     "bouncy": "3.2.2",
-    "browserify": "6.3.3",
-    "browserify-istanbul": "0.1.3",
+    "browserify": "8.1.3",
```
Comment (defunctzombie) [inline package.json]: bumping browserify or other deps should be a separate commit unless this is required to make watchify work

File: lib/control-app.js [hunk 2]
```diff
@@ -5,6 +5,10 @@ var express = require('express');
 var expstate = require('express-state');
 var browserify = require('browserify');
 var im = require('istanbul-middleware');
+var watchify = require('watchify');
+var _ = require('lodash');
```
Comment (defunctzombie) [inline lib/control-app.js]: Can we use the specific lodash modules (https://www.npmjs.com/package/lodash.assign) versus just getting everything. Less likely to break something on an upgrade later.

File: lib/builder-watchify.js [hunk 3]
```diff
@@ -0,0 +1,96 @@
+'use strict';
+
+var _ = require('lodash');
```
Comment (defunctzombie) [inline lib/builder-watchify.js]: same for this lodash: https://www.npmjs.com/package/lodash.where

File: lib/control-app.js (hunk 2 above)
Comment (vvo) [inline lib/control-app.js]: I just stick to what was done in the actual browserify builder for example. Did not want to generate more noise in the PR. Mind doing it (me) in another PR after this one?

Comment (defunctzombie) [inline lib/control-app.js]: yea, that is fine

Comment (vvo) [inline lib/control-app.js]: In fact, I had used lodash specific builds before and stopped because in the end, npm install grabs MORE things, since the 'lodash' package embed everything, while specific builds have a lot of dependencies.

In the end, you might endup with a bigger size of your installed npm package.

What we can do (for code reading) is this: `var assign = require('lodash/object/assign')` which I like a lot. In browserify bundles it will only embed required code (but we are in node here).

File: package.json (hunk 1 above)
Comment (vvo) [inline package.json]: you mean a separate PR, since it's already a separate commit?

Comment (vvo) [inline package.json]: I just did not wanted to lose time with outdated dependencies while using watchify. I know you do not like to update deps for the sake of updating, but waiting too much might do more harm than good?

**Comment by vvo** [comment]
Keep in mind: even if you like this PR (I hope), it will require some cleaning before we merge it of course, we do not want to keep both builders.

File: package.json (hunk 1 above)
Comment (defunctzombie) [inline package.json]: Gotcha. Ok. We will just make a note of it in the change log.

**Comment by defunctzombie** [comment]
I like the PR and feature and don't really see a downside to including it.
If we really want we can have an option to enable disable it. We could even
ship disabled by default at first if we are even more paranoid.

**Comment by vvo** [comment]
I would go for default to watchify and "remove" (it's all there, just faster) browserify builder.

**Comment by vvo** [comment]
Maybe we can wait a little for other devs to think about posible cases where this PR might be a problem

**Comment by vvo** [comment]
So I switched to watchify by default in browserify builder and client-app

**Comment by defunctzombie** [comment]
LGTM.

@rase- ?

**Comment by rase-** [comment]
Looks good.

I think we're going to need to bump lodash though for this to work: `require('lodash/object/assign');`

**Comment by vvo** [comment]
@rase- done, f4a61d23f85b8679c116bf5624743f3fc0d799ae

File: lib/builder-browserify.js [hunk 4]
```diff
@@ -17,7 +22,7 @@ function configure (bundler, cfg) {
     }
 
     function register (type, o) {
-        if ('transform' === type && 'object' === typeof o[type]) {
+        if (type === 'transform' && typeof o[type] === 'object') {
```
Comment (vvo) [inline lib/builder-browserify.js]: Did this to reflect code style from other files in the project

Comment (rase-) [inline lib/builder-browserify.js]: Not really, that's fine.

**Comment by vvo** [comment]
Let's merge and release guys! :)

**Comment by rase-** [comment]
Hmm, getting some hangs in the test suite after this output:

```
  zuul:phantombrowser msg: {"name":"suite :: encountered a declaration exception","type":"test"} +0ms
starting suite :: encountered a declaration exception
```

**Comment by vvo** [comment]
Do you know which test of the test suite fails?

**Comment by vvo** [comment]
Do you know which test of the test suite fails?

**Comment by rase-** [comment]
Looks like something funky is happening both jasmine and qunit tests.

**Comment by rase-** [comment]
And looks like they run fine with `HEAD^^`

**Comment by vvo** [comment]
@rase-, ok gonna have a look

**Comment by vvo** [comment]
Having issues with github and npm right now, will work on this this we.

**Comment by vvo** [comment]
So multiple issues:
- we cannot update lodash because it seems our calls to .where break when using lodash@3. But, it seems there's no need to update lodash.. It just works
- about the "encountered a declaration exception" this is because ATM there's only one possible instance of a builder possible. Like if we create multiple control apps from within the same process (like in our test suite), it fails. I can fix this but it will require a change in the builder API I guess, so that we can have one cache per builder instance, will update PR

File: lib/control-app.js [hunk 5]
```diff
@@ -28,7 +42,7 @@ module.exports = function(config) {
     config.builder = config.builder || '../lib/builder-browserify';
 
     if (config.builder) {
-        build = require(config.builder);
+        build = require(config.builder)();
```
Comment (vvo) [inline lib/control-app.js]: @rase- This is the builder API change, so that multiple instances of control-app can start with different builder "instances". This coupled with the updated browserify-builder, fixes the PR tests

**Comment by vvo** [comment]
@rase- Fixed PR and tests: https://github.com/defunctzombie/zuul/pull/150/files#r25114657

File: lib/builder-browserify.js [hunk 6]
```diff
@@ -25,12 +28,49 @@ function configure (bundler, cfg) {
     }
 }
 
-module.exports = function (files, config, cb) {
+// the builder API is var build = require('builder')(); build(files, config, cb)
+// So that we can have multiple different `watchifyBundler` caches per build instance.
+module.exports = function() {
+    var watchifyBundler;
+
+    return function(files, config, cb) {
```
Comment (vvo) [inline lib/builder-browserify.js]: We could also deal with this using a cache map based on files + config hash but that would require a lot of work without any seen benefit but not breaking the builder API (and custom builders if used?)

**Comment by vvo** [comment]
Reading https://github.com/defunctzombie/zuul/issues/49, breaking the (undocumented) builder option would break the API and we bump major, you tell me.

Also it seems to me that this all boils down to our test suite only, because there should be no way for an external zuul use using only the public API to launch multiple concurrent control-apps in the same node process.

**Comment by vvo** [comment]
tests are ok

**Comment by rase-** [comment]
> we cannot update lodash because it seems our calls to .where break when using lodash@3. But, it seems there's no need to update lodash.. It just works

It doesn't seem to work for me, and it doesn't look like it would work based on how the tag in github for `2.4.1` looks. I get this after `npm install` on the branch:

```
Error: Cannot find module 'lodash/object/assign'
```

**Comment by vvo** [comment]
How the hell does it works on my computer then, magic it seems at first..

I will not use this notation for now.

PR updated

**Comment by rase-** [comment]
Ha, yeah sounds like magic. Maybe there was an `npm link` in place or something?

Tests look good now.

I'm not entirely sure about the builder API change. Thoughts on that @defunctzombie?

**Comment by defunctzombie** [comment]
What is the api change to the builder? Is it backwards compatible?

**Comment by vvo** [comment]
@defunctzombie The api change in the "builder" zuul option is not backward compatible (but I can make it so, read on).

[Before](https://github.com/defunctzombie/zuul/blob/279ab81014ed6b1565144b87fc270fb3db21d210/lib/builder-browserify.js#L28):

```js
module.exports = function (files, config, cb) {
    var opt = {
        debug: true,
        basedir: config.prj_dir
    };
//...
```

[After](https://github.com/vvo/zuul/blob/e694fe24c3c299e6e2b59daa56cc7c8b67c1dc39/lib/builder-browserify.js#L33):

```js
module.exports = function() {
    var watchifyBundler;

    return function(files, config, cb) {
   // ...
```

Now every builder should be declared as a function which when executed, will return the build function.

This is [how we require the builder now](https://github.com/defunctzombie/zuul/pull/150/files#diff-20cb57a5653789fb88cffd4a02f0b95fR45):

```js
var build = require(config.builder)();

// ... later on in an express route:

build(files, config, cb)
```

I had to make this change after discovering that our test suite runs multiple tests with different browserify entry points in the same node process (which is not possible when using the zuul command line, the sole API entry point of zuul). So we cannot have a shared cached (watchify) browserify bundle.

For every new express-app started by zuul in the test suite, we need to have a new cached browserify bundle.

I can make it backward compatible still, by just having an 'if' in our code if we are using the default browserify builder, then we use `var build = require('default-builder')()`. What do you think?

**Comment by vvo** [comment]
On a side note, if we are bumping browserify major, we should bump zuul major since browserify and zuul are tighlty coupled (default builder is browserify).

**Comment by vvo** [comment]
> Ha, yeah sounds like magic. Maybe there was an npm link in place or something?

Yes, in my zuul directory I had done: `npm link` while in another repo that was using lodash, I had done : `npm link zuul`. Sounds like an npm link bug?

**Comment by vvo** [comment]
Modified the PR description to only add the changelog from it

File: lib/builder-browserify.js [hunk 7]
```diff
@@ -17,7 +20,7 @@ function configure (bundler, cfg) {
     }
 
     function register (type, o) {
```
Comment (defunctzombie) [inline lib/builder-browserify.js]: I can't believe I allowed a single letter var name to be merged sometime in the past. I really hope I didn't write that.

**Comment by defunctzombie** [comment]
> if we are bumping browserify major, we should bump zuul major

I really don't want that to be true. Browserify bumps major all the time and is breaking API each time. This is really annoying to keep up with. Browserify should be an implementation detail as much as possible for the general usecases of zuul.

**Comment by defunctzombie** [comment]
@vvo the issue with current builder basically being a global (from what I gather) is indeed not great from an implementation and testing standpoint. Would not mind having that breaking change if it leads to better code.

File: lib/builder-browserify.js (hunk 7 above)
Comment (vvo) [inline lib/builder-browserify.js]: No, I wont say who wrote this :D.

**Comment by vvo** [comment]
@defunctzombie I mislead you, the builder is not a global per se. It's just that if we want to use watchify, we need one cache per express-app which was not possible with the previous builder API.

When exporting a single function as the builder, you cannot have different caches if you have different express app started.

So before changing the builder API, I made my changes backward compatible. PR updated.

About bumping browserify: I thought that since we bump browserify major, we should bump zuul major to reflect this change as it's the default builder.

We can also not bump browserify but in the end, it will mean that we cannot update browserify without bumping major zuul.

**Comment by vvo** [comment]
@defunctzombie @rase- So the PR is ready, backward compatible with the builder option. I advise you to bump major still. Or if you don't, I can downgrade browserify to what it was before.

File: lib/control-app.js [hunk 8]
```diff
@@ -25,10 +41,15 @@ module.exports = function(config) {
 
     var build;
     // default builder is browserify which we provide
-    config.builder = config.builder || '../lib/builder-browserify';
+    config.builder = config.builder || defaultBuilder;
```
Comment (defunctzombie) [inline lib/control-app.js]: @vvo would be better to use the same API for both builders. Maybe I wasn't clear before but I am ok with making a breaking API change for the builder API since it leads to better code architecture for the builder.

Comment (rase-) [inline lib/control-app.js]: Agreed on this.

I'm fine adding another "layer" on the builder API. I don't think many people are using it right now since it's not documented and it should be trivial to wrap any existing builders to the new API.

Looks great otherwise! :)

Comment (vvo) [inline lib/control-app.js]: I just thought about it and it felt like a very specific case of our own builder when running our own test suite. I am not sure it will add any value to anyone else to change the builder API.

Comment (rase-) [inline lib/control-app.js]: Yeah it might not be too useful outside of our needs, but I think it makes sense to keep the API the same across the internal and exposed use cases.

**Comment by vvo** [comment]
Allright, I [updated the PR](https://github.com/vvo/zuul/commit/a64da244372d3364d8243410242a5c001ecd908e#diff-20cb57a5653789fb88cffd4a02f0b95fR46), no prob.

File: lib/control-app.js (hunk 8 above)
Comment (vvo) [inline lib/control-app.js]: Allright, I [updated the PR](https://github.com/vvo/zuul/commit/a64da244372d3364d8243410242a5c001ecd908e#diff-20cb57a5653789fb88cffd4a02f0b95fR46), no prob.

Comment (vvo) [inline lib/control-app.js]: > Looks great otherwise! :)

Thanks, I am using this fork on my projects since some days, you can tell the performance difference when using --local

**Comment by rase-** [comment]
Thanks! :)

**Comment by vvo** [comment]
NICE! Let's bump :-)
